### DISCUSSION TITLE
ESP32 v1.19.1 - pasting long code to REPL is normally broken

### DISCUSSION THREAD

**Discussion by JohnMTu** [micropython/micropython#9470] [General]
Hello,

While testing code from https://github.com/micropython/micropython/discussions/9447#discussioncomment-3763770
i noticed, that with FW "v1.19.1 on 2022-06-18" there is following issue with REPL:
When i try to paste a script to it from:
* intellij (micropython plugin)
* putty (serial connection)

it just fails on random line (36+ usually) or bad indent etc..

Same works fine over Thonny, probably Thonny does some throttling.

But this problem seems not to exist for older firmwares i tested in mentioned discussion.

Any idea why?
thanks

**Comment by jimmo** [comment]
Are you using paste mode? (Ctrl-E)

(And yes, Thonny works in a very different way)

**Comment by JohnMTu** [comment]
yes, reboot, ctrl+E, paste, ctrl+D

**Comment by JohnMTu** [comment]
btw, just tried on v1.11-187-g00e7fe8ab on 2019-08-04; 
there it seems to work for same code as well .. but tried even longer code, and that one failed as well...
actually thinking, if issue might be, that i run it from virtual box.... unfortunatelly cannot try from host atm, no drivers.

**Comment by dpgeorge** [comment]
This is a limitation of using a UART for the REPL.  There is no flow control so the incoming characters come one after the other and at some point the ESP32 cannot keep up with a sustained flow of characters and no rest.

USB serial REPL usually doesn't have this problem because USB has flow control.

**Comment by JohnMTu** [comment]
it was over USB. Board has CH341 converter

**Comment by robert-hh** [comment]
This is effectively a UART communication, because the interface to the ESP32 chip is it's UART link to the CH341 chip, and there is no flow control between the CH341 and the ESP32.

**Comment by karfas** [comment]
I think it's worth to mention that the ESP32 UART **can** handle handshake signals, as well as the CH341 and CP2102 USB/UART converters.
The problem here is that some/many development boards don't connect these signals to the ESP32.
E.g. the ESP32-DevKitC-V4 reference design uses RTS+DTR out of the converter to control the ESP32 program/reset logic on the EN and IO0 pins.

**Comment by Furuidemu** [comment]
WebREPL has the same problem. I have to chop the MicroPython code into pieces and feed them via WebREPL individually. Now I know why.
